### ISSUE TITLE
Prophet 12 module not recognised (Windows 7 / USB)

### ISSUE THREAD

**Issue by Andy2No** [christofmuc/KnobKraft-orm#13]
After choosing auto detect, on the first run after installation, KnobKraftOrm crashed, so I tried to do it from the Setup menu.

Prophet 12 Module was on the list of MIDI devices, so I chose that for the source and destination.  Pressing "Re-check connectivity" then failed - see attached.

I tried it with the MIDI channel set to Omni, and to 1.  Both failed.

I then tried Auto Detect again, which didn't crash this time, but also failed to detect it.  See attached.

**Comment by Andy2No** [comment]
... it seems attaching files in githhub no longer works, with this browser.  Progress.

**Comment by Andy2No** [comment]
... it seems attaching files in githhub no longer works, with this browser.  Progress.

**Comment by Andy2No** [comment]
... it seems attaching files in githhub no longer works, with this browser.  Progress.

**Comment by Andy2No** [comment]
Trying that again, in Chrome - attachments for the first post, above.

![KnobKraft - Prophet 12 module - can't connect and auto detect crashes](https://user-images.githubusercontent.com/8021302/92943341-9010d180-f44a-11ea-9c17-8d4c02652aa0.gif)

![KnobKraft - Prophet 12 module - 2nd auto detect didn't crash, but failed](https://user-images.githubusercontent.com/8021302/92943363-9737df80-f44a-11ea-869f-357b3641a9a5.gif)

**Comment by christofmuc** [comment]
Hi Andy, sorry for the late reply, but I was on vacation. 

To test if the synth is replying at all on the device detect message, can you show the output of the MIDI log screen? There we should see which message the Orm sends out and if and what reply it gets. It did find the USB connection, as we can see "Prophet 12 Module" in the screenshot, so it must be something with the device detect message.

**Comment by Andy2No** [comment]
No problem.

Today, the bug was different, though maybe that was because of running Auto Detect from the MIDI Log page?  This time, it appeared to detect it, but then crashed on the Library page.  

After restarting Orm, the Prophet 12 module wasn't detected again, so I then exited Orm, restarted the module and started Orm again, with roughly the same results.

On the first occasion, Orm crashed as I adjusted the bpm on the module (the flashing bothers me).  On the second, I'd pressed a few buttons on the library page/tab, not touching the synth, then clicked on a browser to try and find out what to do next - since no patches were loaded, and I couldn't see a button to do that.  As I clicked on the task bar, to switch to the browser, Orm crashed for the second time.

Here are some screen shots, numbered in order, with explanatory titles.
![KnobKraft Orm - Prophet 12 module 1 - MIDI Log](https://user-images.githubusercontent.com/8021302/94051411-68692400-fdcf-11ea-8159-82e2de80518b.gif)
![KnobKraft Orm - Prophet 12 module 2 - MIDI Log - Auto detect](https://user-images.githubusercontent.com/8021302/94051416-6901ba80-fdcf-11ea-8a34-e1b4665c833f.gif)
![KnobKraft Orm - Prophet 12 module 3 - crash dialog (Library tab)](https://user-images.githubusercontent.com/8021302/94051417-6901ba80-fdcf-11ea-8e2f-84a7a53bb306.gif)
![KnobKraft Orm - Prophet 12 module 4 -  MIDI Log - after crash](https://user-images.githubusercontent.com/8021302/94051419-699a5100-fdcf-11ea-8858-028a77f49e5c.gif)
![KnobKraft Orm - Prophet 12 module 5 -  MIDI Log - Autodefect after crash](https://user-images.githubusercontent.com/8021302/94051420-699a5100-fdcf-11ea-8046-128d5f82dec2.gif)
![KnobKraft Orm - Prophet 12 module 6 -  MIDI Log - after restarting P12m then Orm](https://user-images.githubusercontent.com/8021302/94051421-6a32e780-fdcf-11ea-8af8-dc246db95385.gif)
![KnobKraft Orm - Prophet 12 module 7 -  MIDI Log - Autodetect  - after restarting P12m then Orm](https://user-images.githubusercontent.com/8021302/94051423-6a32e780-fdcf-11ea-8473-717132d75283.gif)
![KnobKraft Orm - Prophet 12 module 8 -  crashed on Library page again](https://user-images.githubusercontent.com/8021302/94051424-6acb7e00-fdcf-11ea-9d25-6d1b01a9dbe4.gif)

**Comment by Andy2No** [comment]
Incidentally, this time, I can't shut down KnobCraftOrm.exe, either by using it's close button, by using Process Explorer, or from the task manager..  It looks like I may have to restart, to get it to close.

**Comment by Andy2No** [comment]
I hadn't realised github replaces all the file names with gibberish.  They were:

1 - MIDI Log
   (that was after plugging in the P12 module, then starting Orm)
2 - MIDI Log - Auto detect
3 - crash dialog (Library tab)
4 -  MIDI Log - after crash
5 -  MIDI Log - Auto detect after crash
6 -  MIDI Log - after restarting P12m then Orm
7 -  MIDI Log - Auto detect  - after restarting P12m then Orm
8 -  crashed on Library page again

**Comment by christofmuc** [comment]
Ok, that was helpful and surprising. I would think we have (at least) 2 different problems - the non-detection and the crash.

Regarding the non-detection, from the first MIDI log it is pretty clear - the Prophet 12 correctly answers on the request command, but with a different device ID than in the documentation.

The manual states the Prophet 12 should reply with F0 7e 7f 06 02 01 2a 01 00 ..., but your device replies with F0 7e 7f 06 02 01 2b. So this is why the code is not accepting your device as a Prophet 12 - it sends a different message than documented in the manual. I have seen before that DSI is not really consistent there, e.g. the Prophet 8 special edition has a different identifier than the Prophet 8 normal, and maybe in this case they differentiate between the keyboard version and the module, though the module documentation states it should be 2a.

Anyway, we can try to fix this on your computer as a test.

For that, go into the setup page and press the button "Create a new adaption". Select the Prophet 12 and press ok.

This will have created a file called "DSI_Prophet_12.py" in the directory Documents/KnobKraft-Orm-adaptions.

Close the Orm, and open the file in a text editor (Notepad++), we are going to make one change:

 Line 7 should read 

    prophet12_ID = 0b00101011

(notice the 1 at the end, not a 0)

Save the file, and launch the Orm again. The startup log at the bottom should contain a line 
"Overriding built-in adaption DSI Prophet 12 (found in user directory C:\Users\Christof\Documents\KnobKraft-Orm-adaptions)"

With that, the new code with the changed ID is active and ready for a try!

**Comment by Andy2No** [comment]
Thanks.  The folder didn't exist, in Documents, but I found the new .py file here:

C:\Users\andy\KnobKraft-Adaptions

The startup log at the bottom said this:

22:55:53: Overriding built-in adaption DSI Prophet 12 (found in user directory C:\Users\andy\KnobKraft-Adaptions)
22:55:53: Lost communication with DSI Prophet 12 on channel 1 of device Prophet 12 Module - please rerun auto-detect synths!

Incidentally, I use a 1360x768 line screen.  I couldn't see the log, or the buttons at the bottom, until I dragged down the top of the window, then dragged the window up, by the title bar.  I guess that's not a major problem now I know to do it, providing I use it often enough to remember.  It would be good if I didn't have to though.

I'lll post again shortly, when I've tried it with the synth.

**Comment by Andy2No** [comment]
It was looking promising until I pressed the Import Edit Buffer button, which caused it to crash immediately.

It's now refusing to close again, in the same way, but it did eventually yesterday, without restarting, so I'll leave it a while, then try another button.

One thing I should maybe mention is that I have the system font size set larger than standard, at 125%.  I need to do that to make most things in Windows readable to me, but I've noticed it cause problems with some software.

Here's some more screenshots:

1 - MIDI Log
![Orm - P12 module 2b 1 - MIDI Log](https://user-images.githubusercontent.com/8021302/94206630-d5f37e00-febd-11ea-8bf0-5201ab9a3ac1.gif)

2 - MIDI Log - Auto detect
![Orm - P12 module 2b 2 - MIDI Log - Auto detect](https://user-images.githubusercontent.com/8021302/94206632-d68c1480-febd-11ea-9411-f4d584e1e7a8.gif)

3 Library (resized to fit height of small screen - now shows buttons etc at bottom)
![Orm - P12 module 2b 3 Library (resized to fit height of small screen - now shows buttons etc at bottom)](https://user-images.githubusercontent.com/8021302/94206633-d68c1480-febd-11ea-95dd-478d33695106.gif)

4 Library (selected P12)
![Orm - P12 module 2b 4 Library (selected P12)](https://user-images.githubusercontent.com/8021302/94206634-d724ab00-febd-11ea-9e33-443ab1ae5b67.gif)

5 Library - crash on pressing Import edit buffer
![Orm - P12 module 2b 5 Library - crash on pressing Import edit buffer](https://user-images.githubusercontent.com/8021302/94206637-d724ab00-febd-11ea-86cd-683a4620df0f.gif)

**Comment by Andy2No** [comment]
Repeating the process above, after Orm exited and I restarted the Prophet 12 module, the log after auto detect from the MIDI Log screen shows this:

00:00:00.000: Out Prophet 12 Module Sysex [f0 7e 00 06 01 f7]
06:07:00.112: In  Prophet 12 Module Sysex [f0 7e 7f 06 02 01 2b 01 00 00 01 04 02 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
06:07:09.562: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
06:07:10.093: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]

**Comment by Andy2No** [comment]
The smaller window, at the bottom, which I couldn't see without resizing, says this:

23:38:12: Overriding built-in adaption DSI Prophet 12 (found in user directory C:\Users\andy\KnobKraft-Adaptions)
23:38:12: Lost communication with DSI Prophet 12 on channel 1 of device Prophet 12 Module - please rerun auto-detect synths!
23:38:21: No Yamaha Reface DX could be detected - is it turned on?
23:38:22: No Behringer BCR2000 could be detected - is it turned on?
23:38:22: No DSI Prophet 12 could be detected - is it turned on?

**Comment by Andy2No** [comment]
I then tried to re-run autodetect, from the MIDI Log page, but Orm crashed.  It did exit easily though, this time.

![Orm - P12 module 2b 6 crash on re-running auto detect](https://user-images.githubusercontent.com/8021302/94207924-9c704200-fec0-11ea-97bb-1520056c83b0.gif)

**Comment by Andy2No** [comment]
I notice the manual says the device detect message ends with the software version in one byte, but the synth returns two bytes.  I don't know if that matters though:

Manual:
0jjj nnnn Main Software version:   jjj - Minor rev; nnnn - Major rev
followed by F7

Prophet 12 module::
 04 02 f7 - so the message has one extra byte.

I assume this refers to the firmware version, which is shown as 1.4.2.3 - that was also the final release, AFAIK.

**Comment by Andy2No** [comment]
One more thing; I've been using the sysex given for Edit Buffer Dump Request, in the manual, and sending it with Bome's Send SX.  The manual gives the device ID as 2a, so that's what I've been using, and that works...

I've now tried it with 2b, and that also works, so maybe either is acceptable in general.

**Comment by Andy2No** [comment]
Here's the contents of DSI_Prophet_12.py:

#
#   Copyright (c) 2020 Christof Ruch. All rights reserved.
#
#   Dual licensed: Distributed under Affero GPL license by default, an MIT license is available for purchase
#

#prophet12_ID = 0b00101010  # See Page 82 of the prophet 12 manual
prophet12_ID = 0b00101011 # Prophet 12 module returns this

def name():
    return "DSI Prophet 12"


def createDeviceDetectMessage(channel):
    # This is a sysex generic device detect message
    return [0xf0, 0x7e, channel, 0x06, 0x01, 0xf7]


def deviceDetectWaitMilliseconds():
    return 100


def needsChannelSpecificDetection():
    return False


# Prophet 12 Module returns Sysex [f0 7e 7f 06 02 01 2b 01 00 00 01 04 02 f7]

def channelIfValidDeviceResponse(message):
    if (len(message) > 9
            and message[0] == 0xf0  # Sysex
            and message[1] == 0x7e  # Non-realtime
            and message[3] == 0x06  # Device request
            and message[4] == 0x02  # Device request reply
            and message[5] == 0x01  # Sequential / Dave Smith Instruments
            and message[6] == prophet12_ID
            and message[7] == 0x01  # Family MS is 1
            and message[8] == 0x00  # Family member
            and message[9] == 0x00):  # Family member
        # Extract the current MIDI channel from index 2 of the message
        return message[2]
    return -1


def createEditBufferRequest(channel):
    return [0xf0, 0x01, prophet12_ID, 0b00000110, 0xf7]


def isEditBufferDump(message):
    return (len(message) > 3
            and message[0] == 0xf0
            and message[1] == 0x01  # Sequential
            and message[2] == prophet12_ID
            and message[3] == 0b00000011  # Edit Buffer Data
            )


def numberOfBanks():
    return 8


def numberOfPatchesPerBank():
    return 128


def createProgramDumpRequest(channel, patchNo):
    bank = patchNo // numberOfPatchesPerBank()
    program = patchNo % numberOfPatchesPerBank()
    return [0xf0, 0x01, prophet12_ID, 0b00000101, bank, program, 0xf7]


def isSingleProgramDump(message):
    return (len(message) > 3
            and message[0] == 0xf0
            and message[1] == 0x01  # Sequential
            and message[2] == prophet12_ID
            and message[3] == 0b00000010  # Program Data
            )


def nameFromDump(message):
    dataBlock = []
    if isSingleProgramDump(message):
        dataBlock = message[6:-1]
    elif isEditBufferDump(message):
        dataBlock = message[4:-1]
    if len(dataBlock) > 0:
        patchData = unescapeSysex(dataBlock)
        return ''.join([chr(x) for x in patchData[914:931]])
    return "Invalid"


def convertToEditBuffer(channel, message):
    if isEditBufferDump(message):
        return message
    elif isSingleProgramDump(message):
        # Have to strip out bank and program, and set command to edit buffer dump
        return message[0:3] + [0b00000011] + message[6:]
    raise Exception("Neither edit buffer nor program dump - can't be converted")


def unescapeSysex(sysex):
    result = []
    dataIndex = 0
    while dataIndex < len(sysex):
        msbits = sysex[dataIndex]
        dataIndex += 1
        for i in range(7):
            if dataIndex < len(sysex):
                result.append(sysex[dataIndex] | ((msbits & (1 << i)) << (7 - i)))
            dataIndex += 1
    return result

**Comment by Andy2No** [comment]
Looking at edit buffer dumps I've saved, including the one where I sent the ID as 2b instead, of 2a, all dumps contain 2a in the third byte, not 2b.  I assume that's for compatibility with the keyboard version.

I.e. all start with f0 01 2a

I'll check that again later - it's not left set up, when I'm not using it.  I could attach some zipped dumps, of that helps.

**Comment by christofmuc** [comment]
Ok, three ideas:

1. I saw the difference in the documentation with the version number, but that does not matter as I do not use it. So we're safe on that side.
2. If the data dumps that come from the synth always are 2a and not 2b, we could modify the script to keep using 2a everywhere but in the detection also accept 2b. So instead of modifying line 7 as before (revert the change), modify line 34 to something like "and (message[6] == prophet12_ID or message[6]==0x2b)"
3. Our main problem is the crash, because even with wrong data the software should just ignore data that is not detected as being Prophet 12 specific. To test if the software generally works, you could download the syx file with the factory patches from the Sequential website and import it via the "Import sysex file from computer" button, and then click on a patch to see if it gets sent correctly into the synth. If yes, we're looking specifically at a problem with the edit buffer, else the problem is elsewhere.

I assume you have not compiled yourself, but are using a downloaded version? Those contain no debug info, so the dumps won't help, we need to recompile a Debug version or Release With Debug. I can do that in case you don't want to compile yourself.

The dumps then should tell us quickly where the problem is, crashes are normally fixed easily (if you get the line where the crash occurs).

**Comment by Andy2No** [comment]
I'll try that change later.  Thanks.

Yes, I downloaded it from github, just before trying it for the first time, shortly before starting this bug report.  I'm not entirely against installing what's needed to compile it, but I am getting low on hard disk space, and I assume that would be another few gigabytes gone.  A pre-compiled debug version to get started with would be good, anyway.

I've tried loading in the factory sysex, P12_Programs_v1.1c.syx, with "Import sysex files from computer" and got this:

![Orm - P12 module - Import sysex files from computer](https://user-images.githubusercontent.com/8021302/94286900-88265680-ff4d-11ea-85ea-c3357dcbf912.gif)

Are those "Ignoring..." messages to be expected?

**Comment by christofmuc** [comment]
I think you did the #3 before the #2? Sorry I was not clear but the sysex can only be loaded with the fixed file, because it does contain a 2a and not a 2b. 

I will make a Debug version today and that should tell us more about the crash!

**Comment by christofmuc** [comment]
Hi Andy, that was more complex than thought, but I have now implemented proper crash reporting via Internet into the software so we can much more easily find out why your system is crashing. At least that's what I hope!

Also the Fix #2 is already done by me in release 1.7.0, so please give it a try again with your Prophet 12!

Here is the download: https://github.com/christofmuc/KnobKraft-orm/releases/tag/1.7.0

**Comment by christofmuc** [comment]
Hi Andy, I fixed a few things in the crash reporting, maybe skip 1.7.0 and give 1.7.2 a spin, interested to see if that works for you!

**Comment by Andy2No** [comment]
Thanks.  That's a lot better, and hasn't crashed, so far.

This version finds the Prophet 12, and seems to work, but the patch names are not what I expect them to be.  That's a separate bug, or possibly two, so I could start another issue, if you prefer.

Partly, it's hard to check they're right because it's showing the layer B name, but should be showing the layer A name, which is what the synth shows when I change patches.

Partly, some of them are in the wrong order, but it's hard to work out what order they're in.

I'd loaded the FM patches into the first 50 slots, then made some changes to a few of them.  Before that, I'd copied some of the first few patches, which I'd altered, to high numbered slots in bank U1.  That's not what I'm seeing though.

The first patch in Orm still shows as "It's a Prophet! 5" but that seems to be patch 98, which has a "layer A" name of "3 Body Prophet12  +", and a layer B name of "It's a Prophet! 5ths".

The second patch shown is actually patch 99, and is showing the layer B name.  It looks like they all are, and aren't properly in order.  The one shown as patch 64 is patch 62, but there's more to it than that.  It would be easier to check if they were showing the correct patch name, which is the layer A name.

The ones at the end (page 7) appear to be correct, and are the ones at the end of U4, but still show the layer B name (some patches have the same name for both layers so that's not always obvious).

Clicking the patches on Orm loads them into the edit buffer, but the program number isn't changed, so that doesn't help much with working out which is which.

"Import sysex files from computer" appears to work, but the log says they're already in the database, so I can't be sure.

"Import edit buffer from synth" appears to have worked, but I don't know what I can do with it to check.

"Import patches from synth" appears to work, but doesn't seem to do anything different than when I started up Orm, with the Prophet 12 module already connected.

Paging to the end doesn't seem to be consistent (another separate bug, and maybe intermittent).  When I started it up, it loaded less than seven dull pages of patches, with most of the 7th page left blank.  After trying a few things, it apparently had 7 full pages of patches, but I couldn't see the ones I expected to be at the end of bank U4.  

Possibly, it was showing page 6, but indicating that it was page 7, though not letting me page to page 7.

If you can fix the patch name bug, I may be able to work out what else is wrong with the order of the patches.

What I'm seeing would make more sense if I was looking at an older version of bank U1, before I loaded the FM patches.  Is there a cache that needs clearing, maybe?  Perhaps it's showing out of date patch information, from a previous import which worked?  If so, it would have been a couple of weeks ago.  I installed Orm without uninstalling first, though I did remove the custom python file, before running this version.

![Orm 1_7_2 connects but wrong names and or order](https://user-images.githubusercontent.com/8021302/95681900-3c74dd80-0bda-11eb-921b-643ed9a2c9be.gif)

**Comment by Andy2No** [comment]
Okay, I partly get it... maybe.

It says near the top that the patches are imported from P12_Programs_v1.1c.syx 

I didn't tell it to import that file since installing the new version - that's a hangover from the previous version.

So, it seems I may need to delete some state information, to start again.  If you could get it to show the correct patch names, that would help with some of the confusion though.

Also, it appears that importing from the synth isn't working, unless it's skipping most of it because there are already patches loaded when I start it up.  It doesn't just replace what's in memory with what's on the synth, anyway.

**Comment by christofmuc** [comment]
Hi Andy, one way to do a fresh test is to move the database away and start with an empty one.

The database is stored in the directory %APPDATA%\knobkraft (on my machine C:\Users\Christof\AppData\Roaming\knobkraft) and is the file SysexDatabaseOfAllPatches.db3. If you close the Orm and rename this to e.g. db3.backup, on the next start it will create a new database and you can try without the confusing old data.

The order of the patches in the grid is not the same order as in the synth, but rather import data and then program slot in the synth. So it does not make sense to compare the order on the screen with the order in the synth (this is not a bank manager (yet), but rather a database with all your patches no matter where they are in the synth or in some sysex file).

Clicking the patch will send it always into the edit buffer by design, so this is not the way to find out which patch is which. We should fix the formula that extracts the name from the patch - I think I didn't find a documentation on the Sysex format so I just looked into a sysex dump and guessed the name, I might have mistakingly chosen Layer B name. Can you give me an example patch with Layer A name and Layer B name? Could be from the factory programs sysex.

To test "Import edit buffer from synth" just turn a knob, e.g. the cutoff, and it should reimport the changed patch under the same name into a new slot in the grid. There is a diff function to compare to similar named patches and see what's in there (well, only hex data for now but you could see how much two patches differ, for a proper display we'd have to implement all the parameters of the Prophet 12).

Oh, important - it will never delete patches it has seen before, but if you changed patches and reimport, it will create new patches with the exact same name under a different import. In the all patches view, they might be next to each other or completely scattered across the place. The idea is to use the import filter for a specific state, and to hide old versions that you no longer want. I usually don't even bother to hide but rather favorite my patches I want to use and then always look at the patches via the "faves only" filter.

**Comment by Andy2No** [comment]
I see.  Thanks.  Is there a manual I've missed reading?

I've put it away again, but I'll have another look soon.  The first patch in P12_Programs_v1.1c_syx has separate names for the A and B layers, and showed as the B layer name, on the synth: ""It's a Prophet! 5" - truncating "It's a Prophet! 5ths".  The displayed name on the synth is "It's a Prophet!".  

Here's a screenshot of a hex dump, showing the first part of the file, including both layer names (A then B) for that patch, and a zip containing the P12_Programs_v1.1c_syx file.

![P12_Programs_v1 1c_syx first patch hex](https://user-images.githubusercontent.com/8021302/95684155-7ac4c980-0be7-11eb-8a55-8ab83ba1e1f7.gif)

[P12_Programs_v1.1c.zip](https://github.com/christofmuc/KnobKraft-orm/files/5361432/P12_Programs_v1.1c.zip)

**Comment by christofmuc** [comment]
Sadly no, no manual yet. I tried to explain the functionality in the video, but haven't really gotton around to do more videos showcasing the use, certainly need to prioritize that.

So the software seems to work now, the main problem is that it is using Layer B name and not Layer A name, and of course the synth shows Layer A name. I found the error, it can simply be fixed in the Python script. E.g. the patch you show Layer A is called It's a Prophet!, Layer B is called It's a Prophet 5. The way I do this on the Rev2 is to show the name of the patch as "It's a Prophet!|It's a Prophet 5", so you know there are two different layers, and just use one string when layer A and layer B are the same. 

I even added a different separator char to indicate if it is a split, a stack, or two separate patches, but for that I'd need to find the layer mode parameter in the Prophet 12 data. Not sure if it is critically helpful, though. 

So the next release coming soon will have my fix for that, and I promise to do more on documentation!

**Comment by Andy2No** [comment]
Seeing both layer names would be nice, but I was already struggling to read the longer names, where they had a smaller font.  The last screenshot of Orm I gave is full height on my 1360x768 screen, with the scale set to 100%, on the view menu.  I also have system fonts set to 125%, but I don't know if it uses that setting.

There doesn't appear to be room to show any more detail, on a screen like mine, without showing less patches at once, or making them too small to read - unless it's possible to add a scroll bar, or maybe squash the top part more, where there's empty space.  Probably not by much though.

I haven't yet seen it read patches from the synth, and I did try it a few times.  What are you doing to ask it for the dump?  I couldn't find a bank dump request in the Prophet 12 module manual, only individual patch dump requests.   When it's done manually, from the Globals menu on the synth, it shows a count of patches as it does it, which seems to imply it dumps one patch at a time, rather than entire banks.

**Comment by christofmuc** [comment]
I just released version 1.7.3, which shows Layer A name only for the Prophet 12, and has a new scaling factor 75% when it detects a screen with less than 1024 lines, please give that a try! The fonts might get tiny, but were still readable on my screen at 1360x768!

Make sure that your Prophet 12 is detected properly by re-running menu item MIDI... Auto-detect synths and you have a green bar under the Prophet 12 button. In case you have more problems, submit the MIDI log, I added a save log button to the MIDI view.

And you probably want to start fresh, because the new version will not touch the patch names already in your database, so move away the database from %APPDATA%\knobkraft\SysexDatabaseOfAllPatches.db3.

P.S. There is no bank dump request, so the software will basically iterate all patches in a bank and send patch requests.

**Comment by Andy2No** [comment]
I'm not clear if replying to a closed issue works, but this seems to belong here.

In V1.7.3, the patch names from an imported sysex file are fine, but I can't get Orm to detect the P12 module, again, so I can't currently test anything else.  Here is the MIDI log:


00:00:00.000: Out Prophet 12 Module Sysex [f0 7e 00 06 01 f7]
00:42:06.448: In  Prophet 12 Module Sysex [f0 7e 7f 06 02 01 2b 01 00 00 01 04 02 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:42:45.373: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:42:45.919: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]

Possibly I can get around it by creating a new variation, and editing that again.  I'll look into that soon.

**Comment by christofmuc** [comment]
On which MIDI channel is the Prophet12? The return value (second line in the file) lists 0x7f in the byte reserved for the MIDI channel, that would be invalid. Is it in Omni mode? 

I do reopen this issue.

**Comment by christofmuc** [comment]
Documentation says this indicates Omni mode - please switch it to any other channel (turn off Omni) and see if it works then!

**Comment by Andy2No** [comment]
After setting it to MIDI channel 1, in Globals, I still get this:

00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
01:17:43.892: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
01:17:44.444: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]


22:37:57: No DSI Prophet 12 could be detected - is it turned on?


I neglected to say, I can still click on an imported patch name, in the library, and it loads into the edit buffer of the P12 module, so that part still works.

**Comment by Andy2No** [comment]
After also setting it in Orm's settings page, for the P12, I get the same behaviour:

00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
01:20:44.820: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
01:20:45.347: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]


22:40:58: No DSI Prophet 12 could be detected - is it turned on?

**Comment by Andy2No** [comment]
After turning both off and back on again (module first), but NOT running auto-detect, I get this:

00:00:00.000: Out Prophet 12 Module Sysex [f0 7e 00 06 01 f7]
01:23:22.437: In  Prophet 12 Module Sysex [f0 7e 00 06 02 01 2b 01 00 00 01 04 02 f7]


22:43:35: Overriding built-in adaption DSI Prophet 12 (found in user directory C:\Users\andy\KnobKraft-Adaptions)

(I made the adaptation to look at, a short while ago, but haven't edited it)

I notice I don't see a response from the module, in the last couple of auto-detect logs.

I'll try it as it is currently, without running auto-detect, and see what else is working.

**Comment by christofmuc** [comment]
I made a 1.7.4 with a quick fix for the Omni mode, but for that you'd need to delete your overriding adaption.

Normally, you don't need to set anything in the Setup tab except choosing which synths you want to use. The MIDI device and channel is filled out automatically when running the MIDI... Auto-detect synths function (if it works correctly, as it should *now*). 

Then the bar under the synth name will turn green to indicate you have a successful detection and a running bidrectional communication with the device.

On restart, the Orm only does a quick detect - it sees if the device is still on the same port and channel as it was the last time it was used. If that is the case, you get the green bar, else the best way to fix it is to run MIDI... Auto-detect synths again. Hope this makes sense.

**Comment by Andy2No** [comment]
Okay, Import Edit buffer is working, and seems fine.

Import patches from synth, stalls at 78% complete on the progress bar, but I didn't get the menu to do that in the previous version, so that's promising.

This ought to be a separate issue, but here's the log of trying to import bank 1.  I also tried bank 5, with the same result:

00:00:00.000: Out Prophet 12 Module Sysex [f0 01 2a 05 00 00 f7]
01:40:41.695: In  Prophet 12 Module Sysex [f0 01 2a 02 00 00 00 30 24 46 30 32 32 32 00 32 49 00 00 08 04 04 00 05 0d 00 01 02 03 07 00 08 09 0a 04 05 06 07 00 40 40 40 40 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 01 01 01 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 07 07 00 00 00 00 00 00 00 00 00 00 00 15 00 05 00 00 00 60 00 15 7f 7f 00 2a 53 00 50 3c 07 09 05 03 00 00 00 00 00 56 2e 00 00 00 00 00 00 00 7f 7f 7f 00 7f 00 00 00 00 00 00 04 00 00 06 54 54 2a 00 00 00 00 00 00 00 03 00 00 00 28 0f 07 05 29 29 00 32 00 00 00 00 00 00 00 00 00 00 00 25 00 00 00 00 00 00 7f 00 40 40 0d 14 7f 7e 11 58 7f 00 00 2e 00 00 00 00 00 00 00 00 00 57 66 7f 18 2e 00 00 1c 0d 2f 1a 0f 36 00 07 11 00 00 00 00 00 00 00 00 00 0e 0f 10 13 00 0b 0c 17 06 0d 0e 00 60 00 00 00 00 00 71 70 1d 20 7f 7d 7e 5a 7f 7f 01 4d 7f 7f 7f 7f 7f 7f 00 2d 24 26 23 07 09 07 00 29 00 2e 00 00 00 00 00 00 00 00 00 78 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 03 00 02 00 00 00 06 00 00 40 01 00 00 00 00 3c 3c 7f 3c 3c 3c 3c 3c 3c 3c 7f 3c 3c 3c 3c 3c 3c 3c 7f 3c 3c 3c 3c 3c 3c 3c 7f 3c 3c 3c 3c 3c 3c 3c 03 3c 3c 78 78 78 78 78 00 78 78 78 78 78 78 78 00 78 78 78 78 78 78 78 00 78 78 78 78 78 78 78 00 78 78 78 78 78 78 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 20 46 4d 20 00 20 52 68 6f 64 65 73 00 20 20 31 20 20 20 20 00 20 20 00 00 3c 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0e 24 24 0e 32 32 00 32 32 70 69 69 6a 0b 00 0f 06 05 00 01 02 03 00 07 08 09 0a 04 05 06 20 07 40 40 40 40 26 00 00 4a 00 00 00 00 00 22 00 00 00 00 00 00 00 00 00 46 46 46 46 00 00 00 00 00 01 01 01 01 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 07 07 00 00 00 00 00 00 00 00 00 00 00 2d 00 00 01 00 00 00 00 15 6f 7f 00 00 00 48 50 3c 07 09 05 03 00 00 00 00 00 05 00 00 00 00 00 00 00 00 7f 7f 00 7f 7f 00 00 00 00 00 78 00 00 00 06 06 06 06 00 00 00 00 00 00 00 00 00 00 00 00 00 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7f 00 40 00 40 7a 7f 7f 7f 7f 7f 00 00 00 00 00 00 00 00 00 00 00 00 0e 58 36 32 ...

That does look like patch dumps, so maybe it's trying to retrieve too many patches?  They're numbers 1-99 (oddly), in each bank.  Is it trying to get a bank of 128, and waiting for ever for the one hundredth patch, maybe?

Running auto-detect after that, fails again, like this:

00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
01:46:10.002: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
01:46:10.556: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]

23:06:24: No DSI Prophet 12 could be detected - is it turned on?

So, the detection on startup works okay, at least it did this time, but the auto-detect didn't.

Despite the failure to auto-detect, I can still do the same things as before - import the edit buffer, or try to import patches (though it gets stuck at 78%).

So, I'll just ignore the failure messages for now.

**Comment by Andy2No** [comment]
Changing this:

def numberOfPatchesPerBank():
    return 128


to 

def numberOfPatchesPerBank():
    return 99

works!

**Comment by christofmuc** [comment]
Ok, the module seems to be easily irritated by sysex transfers, but I have heard bad things about the quality of Sequential's MIDI implementation, and my OB-6 is full of bugs.

Good news is - yes, if the Prophet 12 has only 100 patches per bank, that needs to be changed in line 63 of the adaption code, it should be return 99 instead of return 128, you can do that change yourself (requires restart of Orm). I'll fix it for the next release.
Why does it only have 99 patches per bank and not 100? well...

You can also try to increase the timeout value in line 20, currently I assume the Prophet is fast and replies within 100 ms as does my Rev2, but maybe if auto detection is flaky give it a try and increase this to 200.

**Comment by christofmuc** [comment]
> 
> 
> Changing this:
> 
> def numberOfPatchesPerBank():
> return 128
> 
> to
> 
> def numberOfPatchesPerBank():
> return 99
> 
> works!
Uff, brilliant! Getting there!

**Comment by Andy2No** [comment]
Yep :)

I hadn't seen your other messages, so that was still with 1.7.3.  I'll try 1.7.4, with the MIDI channels set back to All (Omni) on the module, after a short break.  Thanks.

One thing that bugs me, from the manual, is it says not to use DIN MIDI In and USB MIDI at the same time - which suggests the implementation is quite fragile.  I'd normally want a keyboard or wind controller connected to DIN MIDI In, and the PC connected to USB MIDI, for patch dumps, or Orm.

I've wondered whether I need to make a MIDI event processor to filter out ActiveSense messages from the keyboard / controller, to stop them interrupting sysex messages over USB.  So far, I've just been turning the keyboard off, when I need to use USB.  Today, I've been trying it with Virtual MIDI Keyboard instead (a free on-screen keyboard) via one of the MIDI interfaces plugged into my PC, which has been okay.  I doubt if that sends ActiveSense, but I'm not sure.

They may have fixed that, of course, since the manual was written, but I haven't seen any mention of it.

I did try auto-detect without the cable to DIN MIDI In too, but it was the same (in 1.7.3).

**Comment by Andy2No** [comment]
After deleting the adaptation, then Immediately after installing V1.7.4, with the Prophet 12 module already turned on and connected, with the MIDI channel set to ALL, the log says:

00:47:30: Lost communication with DSI Prophet 12 on channel 1 of device Prophet 12 Module - please rerun auto-detect synths!

At this stage, I can send a patch from the library to the edit buffer, but apparently can't do anything else.

Running auto-detect then gives this:

00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
03:29:47.807: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
03:29:48.358: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]


00:33:15: No DSI Prophet 12 could be detected - is it turned on?

Turning both off and back on again (module first) gives the same startup log:

00:52:37: Lost communication with DSI Prophet 12 on channel 1 of device Prophet 12 Module - please rerun auto-detect synths!

And the same, or similar, result from running auto-detect:

00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
03:16:23.341: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
03:16:23.891: In  LoopBe Internal MIDI Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 00 20 32 7f 15 01 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 00 20 32 7f 15 01 f7]

00:53:25: No DSI Prophet 12 could be detected - is it turned on?



Possibly, I just got lucky with the connection on startup working, in V1.7.3?  Perhaps it's intermittent.

I also get the same result if I manually set the channel number to 1 on the synth... and if I also then set it to 1 on the Settings page of Orm.

However!  Having done the manual setting to CH1 (possibly unrelated), and restarting Orm again, it has detected the module when it started up, and I can import the edit buffer, and a bank of patches.

00:56:14: Detected DSI Prophet 12 on channel 1 of device Prophet 12 Module

Auto-detection still fails thuogh, in the same way:

00:00:00.000: Out Microsoft GS Wavetable Synth Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
03:37:27.492: In  LoopBe Internal MIDI Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out MIDISPORT Uno Out Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out A Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out B Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out C Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out 5- MIDISPORT 4x4 Out D Sysex [f0 43 20 7f 1c 05 00 00 00 f7]
00:00:00.000: Out Prophet 12 Module Sysex [f0 43 20 7f 1c 05 00 00 00 f7]

Prophet 12 could be detected - is it turned on?


I should probably go to bed but I'll experiment with anything else you want to suggest, tomorrow.

It seems, at the moment, the thing to to still, is set the channel manually, on both, then it finds it on startup, but still not from auto-detect.  Whether it finds it consistently, on startup, it's too early to say.

Incidentally, if I disable all other synths, on the Settings page, I don't see anything on the MIDI log when I try to do an auto-detect (separate bug, and not really important, right now).

**Comment by christofmuc** [comment]
I made a 1.7.5 which includes the change to the 99 patches per bank so it doesn't stop at 78%.

So I think the detection code is correct now - we need to look at your connectivity of the Prophet 12.

The manual of the keyboard Prophet 12 states on USB:
"Note: MIDI In and USB should not be used at the same time, as
overlapping messages from different sources may cause the Prophet
12 to respond unpredictably"

I think you said so before. I fear we need to take this seriously, as we can see from the logs that the Prophet 12 is not responding to a correctly send device enquiry message. Can you test with everything MIDI disconnected except the Prophet 12 USB? That should work - if not, there is only the timeout we can adjust (see above). 

The USB stuff seems to be bit more fragile, as the manual continues:

"In Windows, if you unplug the USB cable and plug it back in while a
program has the Prophet 12 port open, you may have to resync. That
usually means going to the Prophet 12 Keyboard Properties—in the
Windows Device Manager under “Sound, video, and game controllers”—
and clicking OK"

I'd like to get you onto stable ground, because normally this should work without a hitch. The manual setup is more a fallback for old synths that can not be detected. The critical thing to look for in the MIDI log is this line:

"00:42:06.448: In Prophet 12 Module Sysex [f0 7e 7f 06 02 01 2b 01 00 00 01 04 02 f7]"

Now that we fixed the Omni mode (the third byte being 7f), the Orm should detect it allright, if the Prophet 12 replies with a message like above. If it doesn't reply, it is in my eyes in a half-disconnected state, and the only way to fix that would be turn on/off of the Prophet 12 - and in the case of USB probably also restarting the Orm and worst case also going through the Windows procedure from the manual above. 

BTW, Sequential must have fixed something with the newer synths, as my Rev2 has a setup for "Sysex MIDI", and I can select normal MIDI to go to DIN and all Sysex to go to USB, which is the setup that makes most sense I think. 

[MIDI log bug - yes, the first line of the log is not shown / scrolled out of view]

**Comment by christofmuc** [comment]
I assume this works, closing for now. In case you have problems with the autodetection - test with isolated devices. E.g. if my MKS80 is turned on and sends out all kinds of messages, the auto detection can become a bit random. Version 1.8.0 of the KnobKraft Orm will have an improved setup screen where manual entry is easier than now.

**Comment by Andy2No** [comment]
Sorry for the slow reply.  I've been a bit distracted, lately.  I haven't yet tried it with the latest version and only a USB connection, with a longer timeout, and so on.

Although it might work without the DIN MIDI connection in use, I wouldn't have much use for it like that.  I guess I could use a merger box, and connect both Orm and the keyboard that way, if I had to, but it would be less than ideal.

Maybe it would be better to wait to try it with the new version, when it's ready.

**Comment by Andy2No** [comment]
The MIDI log bug I've encountered is that the auto detect process isn't shown on the log at all, if I disable all of the other synths on the settings page, by unticking them.

I previously had the Reface DX and Behringer BCR2000 left enabled in the settings, because I own those, though I haven't tried using them with Orm yet.  

Since Orm always started with one of those selected (I forget which), but not the Prophet 12 module, I tried disabling them, so there was just the Prophet 12 module left, but it meant that when I asked it to autodetect, while looking at the MIDI log, nothing at all appeared in the log.

**Comment by christofmuc** [comment]
Oh, don't worry! 

Regarding the DIN MIDI - yes, at some point when my setup grew I decided to go full DIN MIDI with a proper MIDI patch bay, I use a Motu Midi Express XT. It's much slower for patch download, but for everything else the setup is more robust now and I won't go back to USB. 

The MIDI log bug I couldn't reproduce with the latest development build, maybe I accidentally fixed it or it is again a USB topic. If this comes up anywhere, I'll open a new ticket for that one because it seems unrelated to the Prophet 12 itself.
